File: backend/app/models/assignment.py
"""
VEP MVP Backend - Assignment Models

Database models and Pydantic schemas for assignments and assignment_voters.
"""

from datetime import date, datetime
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import Field, SQLModel


class AssignmentStatus:
    """Assignment status constants."""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class AssignmentBase(SQLModel):
    """Base assignment fields shared across schemas."""
    name: str
    description: Optional[str] = None
    assigned_date: date = Field(default_factory=date.today)
    due_date: Optional[date] = None
    status: str = Field(default=AssignmentStatus.PENDING)


class Assignment(AssignmentBase, table=True):
    """
    Assignment database model.
    
    Stores canvassing assignments linking managers to canvassers.
    Status lifecycle: pending → in_progress → completed (or cancelled).
    """
    __tablename__ = "assignments"
    
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    user_id: UUID = Field(foreign_key="users.id")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class AssignmentVoter(SQLModel, table=True):
    """
    AssignmentVoter junction table model.
    
    Links assignments to voters (many-to-many relationship).
    Allows pre-planned route optimization via sequence_order.
    """
    __tablename__ = "assignment_voters"
    
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    assignment_id: UUID = Field(foreign_key="assignments.id")
    voter_id: UUID = Field(foreign_key="voters.id")
    sequence_order: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)


class AssignmentCreate(AssignmentBase):
    """Schema for creating a new assignment."""
    user_id: UUID
    voter_ids: list[UUID] = Field(default_factory=list)


class AssignmentUpdate(SQLModel):
    """Schema for updating an assignment."""
    name: Optional[str] = None
    description: Optional[str] = None
    due_date: Optional[date] = None
    status: Optional[str] = None


class AssignmentRead(AssignmentBase):
    """Schema for reading an assignment."""
    id: UUID
    user_id: UUID
    voter_count: int = 0
    completed_count: int = 0
    created_at: datetime
    updated_at: datetime


class AssignmentWithVoters(AssignmentRead):
    """Schema for reading an assignment with voter details."""
    voters: list = Field(default_factory=list)

File: ios/VEP/Models/Assignment.swift
import Foundation

/// Assignment model representing a canvassing assignment
struct Assignment: Codable, Identifiable {
    let id: UUID
    let name: String
    let description: String?
    let assignedDate: Date
    let dueDate: Date?
    let status: AssignmentStatus
    let voterCount: Int
    let completedCount: Int
    var voters: [Voter]?
    
    enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case assignedDate = "assigned_date"
        case dueDate = "due_date"
        case status
        case voterCount = "voter_count"
        case completedCount = "completed_count"
        case voters
    }
    
    /// Progress percentage (0.0 to 1.0)
    var progress: Double {
        guard voterCount > 0 else { return 0.0 }
        return Double(completedCount) / Double(voterCount)
    }
    
    /// Whether the assignment is overdue
    var isOverdue: Bool {
        guard let dueDate = dueDate else { return false }
        return dueDate < Date() && status != .completed
    }
}

/// Assignment status enumeration
enum AssignmentStatus: String, Codable {
    case pending
    case inProgress = "in_progress"
    case completed
    case cancelled
    
    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .inProgress: return "In Progress"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        }
    }
}
